### ISSUE TITLE
TeachingTip: HeroContentPlacement API has several issues

### ISSUE THREAD

**Issue by Felix-Dev** [microsoft/microsoft-ui-xaml#3254]
**Describe the bug**
The [HeroContentPlacement](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.teachingtipherocontentplacementmode) API of the `TeachingTip` control has a few issues. The documentation for it states that it is responsible for positioning the hero content inside the TeachingTip. So If I set a `HeroContentPlacement` of `Bottom` then the hero content should be displayed at the bottom of the TeachingTip.

If I want to change the position of the TeachingTip relative to its target, for example, then I have to use the `PreferredPlacement` API. By default, the PreferredPlacement is set to `Auto` which tries to place the TeachingTip above its target control.

So given the following XAML markup
```xml
<Button Content="Show TeachingTip">
    <Button.Resources>
        <muxc:TeachingTip x:Name="ToggleThemeTeachingTip1"
                          Target="{Binding ElementName=TeachingTipXAMLMarkupButton}"
                          Title="Title" Subtitle="Subtitle"
                          HeroContentPlacement="Bottom">
            <muxc:TeachingTip.HeroContent>
                <Rectangle Fill="Green" MinHeight="100" MinWidth="100"/>
            </muxc:TeachingTip.HeroContent>
        </muxc:TeachingTip>
    </Button.Resources>
</Button>
```
I expect the TeachingTip to be placed above the button, with its hero content placed at the bottom of the TeachingTip.

Yet this is what I get as a result:
![image](https://user-images.githubusercontent.com/1398851/92336032-c3470f80-f09c-11ea-9870-803106ff8aff.png)

We are seeing two issues here:
1) The TeachingTip is placed below the button even though it should be placed above it.
2) The hero content is the top content of the TeachingTip even though it should be the bottom content.

If I change the `HeroContentPlacement` to `Top` in my XAML above, I get this result:
![image](https://user-images.githubusercontent.com/1398851/92333407-90455180-f085-11ea-8f8b-4f1fb54f48bd.png)

While the hero content is now correctly set as the top content of the TeachingTip, observe that the TeachingTip has also changed its position relative to the button - it is now placed above it! In other words, changing `HeroContentPlacement` from Top to Bottom moves the TeachingTip below its target, yet keeps the hero content as the top content of the TeachingTip. This is wrong. The TeachingTip should still be above its target, yet its hero content should be the bottom content of the TeachingTip. Changing the value of the `HeroContentPlacement` property should not affect the position of the TeachingTip relative to its target (`PreferredPlacement` exists for that).

And indeed, looking at the TeachingTip implementation, we are seeing that the `HeroContentPlacement` affects the placement of the TeachingTip related to its target: https://github.com/microsoft/microsoft-ui-xaml/blob/a1545320d01454cd3d166fac3e9fbe8dd0b68201/dev/TeachingTip/TeachingTip.cpp#L1743

So to summarize, we have two bugs for the `HeroContentPlacement` API here:
* Bug 1: It affects the placement of the *TeachingTip* even though it should only affect the placement of the *hero content* inside the TeachingTip. It overrides the setting of the `PreferredPlacement` API which should solely be responsible for positioning the TeachingTip in relation to its target.
* Bug 2: Explicitly setting it to `Bottom` does not position the hero content at the bottom of the TeachingTip as it should.

**Expected behavior**
```xml
<Button Content="Show TeachingTip">
    <Button.Resources>
        <muxc:TeachingTip x:Name="ToggleThemeTeachingTip1"
                          Target="{Binding ElementName=TeachingTipXAMLMarkupButton}"
                          Title="Title" Subtitle="Subtitle"
                          HeroContentPlacement="Bottom">
            <muxc:TeachingTip.HeroContent>
                <Rectangle Fill="Green" MinHeight="100" MinWidth="100"/>
            </muxc:TeachingTip.HeroContent>
        </muxc:TeachingTip>
    </Button.Resources>
</Button>
```
![image](https://user-images.githubusercontent.com/1398851/92334324-0c8f6300-f08d-11ea-86ce-4de64770a589.png)

**Version Info**

NuGet package version: 
Microsoft.UI.Xaml 2.4 and earlier
Microsoft.UI.Xaml 2.5 previews

Windows app type:
| UWP              | Win32            |
| :--------------- | :--------------- |
|  |  |

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)              |    |
| May 2020 Update (19041)            | Yes   |
| November 2019 Update (18363)       |    |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            |  |
| Xbox               |  |
| Surface Hub        |  |
| IoT                |  |

**Additional Contect**
I would like to look into this.

**Comment by StephenLPeters** [comment]
This is very strange, the MUXControls test app does not behave this way.  There is a design requirement that for auto placement mode we do not open in a position which the pointer extends from the hero content. Which means when you set your HeroContentPlacement to bottom the auto placement is going to prefer bottom instead of top, which you can see in the mux controls test app. However I don't see how you got the hero content to show up on top, very strange.

**Comment by Felix-Dev** [comment]
@StephenLPeters 

> This is very strange, the MUXControls test app does not behave this way. 

I can reproduce this in the MUXControls test app: (WinUI master build)
```xml
<Button x:Name="TeachingTipExampleButton" Content="Show tip"
        Click="Button_Click">
    <Button.Resources>
        <muxc:TeachingTip x:Name="ButtonTeachingTip" Title="Title" Subtitle="Subtitle"
                          Target="{Binding ElementName=TeachingTipExampleButton}"
                          HeroContentPlacement="Bottom">
            <muxc:TeachingTip.HeroContent>
                <Rectangle Fill="ForestGreen" MinHeight="100" MinWidth="100" />
            </muxc:TeachingTip.HeroContent>
        </muxc:TeachingTip>
    </Button.Resources>
</Button>
```

```cs
private void Button_Click(object sender, Windows.UI.Xaml.RoutedEventArgs e)
{
    this.ButtonTeachingTip.IsOpen = true;
}
```
![image](https://user-images.githubusercontent.com/1398851/92619615-08985680-f2c2-11ea-884c-475902864e34.png)


> There is a design requirement that for auto placement mode we do not open in a position which the pointer extends from the hero content. Which means when you set your HeroContentPlacement to bottom the auto placement is going to prefer bottom instead of top

Either the implementation is wrong or the documentation. According to the documentation, the [PreferredPlacement](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.teachingtip.preferredplacement?view=winui-2.4#Microsoft_UI_Xaml_Controls_TeachingTip_PreferredPlacement) API only controls the position of TeachingTip relative to its target (or parent window). The placement of the hero content plays no role here. 
Looking at the documentation for the [HeroContentPlacement](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.teachingtip.herocontentplacement?view=winui-2.4#Microsoft_UI_Xaml_Controls_TeachingTip_HeroContentPlacement) API, it is the sole API responsible for the placement of the hero content inside the TeachingTip. If I do no want the TeachingTip tail extending from the hero content, then I have to set the `HeroContentPlacement` to Auto. However, if I set it to either `Top` or `Bottom` then the hero content should be placed at the top or bottom of the TeachingTip regardless of the tail position.

In short, the documentation does not state that setting the `PreferredPlacement` of the TeachingTip has any effect on the position of the hero content inside the TeachingTip. The [implementation](https://github.com/microsoft/microsoft-ui-xaml/blob/a1545320d01454cd3d166fac3e9fbe8dd0b68201/dev/TeachingTip/TeachingTip.cpp#L1743) makes such an assumption though. 

So given my XAML above, I expect the TeachingTip to be placed *above* the button (as `PreferredPlacement` is Auto) and the the HeroContent should be placed at the *bottom* of the TeachingTip as `HeroContentPlacement` is Bottom).

Either the documentation or the implementation is not correct here.

**Comment by StephenLPeters** [comment]
Does the repro have something to do with setting those properties from markup? because setting those same properties using the test pages combobox's does not reproduce the issue.

The preferred placement does (should) not affect the HeroContentPlacement (unless HeroContentPlacement is auto). However HeroContentPlacement does affect the teaching tip placement. The PreferredPlacement is called preferred because we don't guarantee that we will honor it, the two additional aspects we check are that the tip will fit within the app and that the tail does not protrude from the hero content.

**Comment by Felix-Dev** [comment]
@StephenLPeters 

>  However HeroContentPlacement does affect the teaching tip placement.

This is not what the [documentation](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.teachingtipherocontentplacementmode?view=winui-2.4) is saying though. So the documentation needs to be updated here if that is desired behavior.

> Does the repro have something to do with setting those properties from markup? because setting those same properties using the test pages combobox's does not reproduce the issue.

I haven't checked the TeachingTip testing page. I have instead created my own teaching tip using the XAML markup above. Setting HeroContentPlacement from markup is an issue because if I set it to Bottom then the [corresponding VisualState](                                <VisualState x:Name="HeroContentBottom">) won't be entered (I pointed that out in my opening post above as one of the two issues I am seeeing).

**Comment by StephenLPeters** [comment]
@SavoySchuler Can you weigh in on the pointer  protruding from the hero content? I agree that this is confusing, I feel like we had our reasons previously, and those reasons should be documented.

@Felix-Dev if setting heroContentPlacement from markup doesn't cause the controls VSM to go to the appropriate states (like it does when you set the property from code) then that is certainly a bug.

**Comment by Felix-Dev** [comment]
Here is how I would expect the `PreferredPlacement` and `HeroContentPlacement` APIs to behave:
* The position of the TeachingTip should only be controlled by the `PreferredPlacement` API
* The position of the hero content is only controlled by the `HeroContentPlacement` API:
  * If I want my hero content to be placed at the top of the TeachingTip (regardless of the tail position) then I set it to Top
  * If I want my hero content to be placed at the bottom of the TeachingTip (regardless of the tail position) then I set it to Bottom
  * If I want my hero content to be placed "opposite" of the tail, then I set it to Auto. So if the TeachingTip's placement logic determines that the TeachingTip is placed above the target, then my hero content should be the top content of the tip. Likewise, if the the tip's placement is below the target then the hero content should be the bottom content of the tip. 

So while the `HeroContentPlacement` would not affect the psoition of the TeachignTip on the screen, if I set it to Auto, then it would base the position of the hero content on the determined position of the TeachingTip.

@StephenLPeters Since these are two issues, I will create a search issue for "HeroContentPlacement ignored from XAML markup" and we can keep using this to discuss, if and what should be changed about the interplay of the `PreferredPlacement` and `HeroContentPlacement` APIs.

**Comment by Felix-Dev** [comment]
@SavoySchuler Ping for your thoughts on the apparent mismatch between the documentation and the actual behavior of the `HeroContentPlacement` API in regards to affecting the teaching tip's placement.
